lib/services/semaphore.go

```diff
@@ -0,0 +1,604 @@
+}
+
+// Match checks if the supplied semaphore matches this filter.
+func (f *SemaphoreFilter) Match(sem Semaphore) bool {
+	if f.SemaphoreKind != "" && f.SemaphoreKind != sem.GetSubKind() {
+		return false
+	}
+	if f.SemaphoreName != "" && f.SemaphoreName != sem.GetName() {
+		return false
+	}
+	return true
+}
+
+type SemaphoreLockConfig struct {
+	// Service is the service against which all semaphore
+	// operations are performed.
+	Service Semaphores
+	// Expiry is an optional lease expiry parameter.
+	Expiry time.Duration
+	// Params holds the semaphore lease acquisition parameters.
+	Params AcquireSemaphoreParams
+}
+
+// CheckAndSetDefaults checks and sets default parameters
+func (l *SemaphoreLockConfig) CheckAndSetDefaults() error {
+	if l.Service == nil {
+		return trace.BadParameter("missing semaphore service")
+	}
+	if l.Expiry == 0 {
+		l.Expiry = defaults.SessionControlTimeout
+	}
+	if l.Expiry < time.Millisecond {
+		return trace.BadParameter("sub-millisecond lease expiry is not supported: %v", l.Expiry)
+	}
+	if l.Params.Expires.IsZero() {
+		l.Params.Expires = time.Now().UTC().Add(l.Expiry)
+	}
+	if err := l.Params.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// SemaphoreLock provides a convenient interface for managing
+// semaphore lease keepalive operations.
+type SemaphoreLock struct {
+	cfg      SemaphoreLockConfig
+	cancel   context.CancelFunc
+	doneC    <-chan struct{}
+	renewalC chan struct{}
+	cond     *sync.Cond
+	err      error
+	fin      bool
+}
+
+// finish registers the final result of the background
+// goroutine.  must be called even if err is nil in
+// order to wake any goroutines waiting on the error
+// and mark the lock as finished.
+func (l *SemaphoreLock) finish(err error) {
+	l.cond.L.Lock()
+	defer l.cond.L.Unlock()
+	l.err = err
+	l.fin = true
+	l.cond.Broadcast()
+}
+
+// Done signals that lease keepalive operations
+// have stopped.
+func (l *SemaphoreLock) Done() <-chan struct{} {
+	return l.doneC
+}
+
+// Error blocks until the final result is available.  Note that
+// this method may block longer than desired since cancellation of
+// the parent context triggers the *start* of the release operation.
+func (l *SemaphoreLock) Error() error {
+	l.cond.L.Lock()
+	defer l.cond.L.Unlock()
+	for !l.fin {
+		l.cond.Wait()
+	}
+	return l.err
+}
+
+// Cancel cancels associated lease keepalive.
+func (l *SemaphoreLock) Cancel() {
+	l.cancel()
+}
+
+// Renewed notifies on next successful lease keepalive.
+// Used in tests to block until next renewal.
+func (l *SemaphoreLock) Renewed() <-chan struct{} {
+	return l.renewalC
+}
+
+// AcquireSemaphoreLock attempts to acquire and hold a semaphore lease.  If successfully acquired,
+// background keepalive processes are started and an associated lock handle is returned.
+func AcquireSemaphoreLock(ctx context.Context, cfg SemaphoreLockConfig) (*SemaphoreLock, error) {
```

Comment:
I think this function should live in `lib/client/semaphore.go` and be callable from the Auth Server client. We don't have a whole lot of application logic that sits in `lib/services`, it's mainly data structures, so it feels out of place. @fspmarshall what do you think?

Reply:
Good point.  Unfortunately, it can't go in `client` due to circular import issues (`client` depends on `auth` depends on `services/suite`  depends on `AcquireSemaphoreLock`).  I kept it in `services` because keeping semaphore stuff together seemed like a reasonable default, but I'm open to alternatives.

Reply:
Can you use an interface to break the circular dependencies?

Reply:
Maybe... I could make a `SemaphoreLock` interface and an `AcquireSemaphoreLockFunc` type and move the concrete impls somewhere else.  `SemaphoreLockConfig` would still need to be defined in services, and wherever we moved the concrete implementation, it would still need to be importable  from `services/suite`.  We could put it in `services/local` if we made `services/suite` accept a `SemaphoreLockFunc` parameter, but thats a lot of indirection for something that doesn't actually need to have multiple concrete implementations.

I think `SemaphoreLock` falls into the same category as things like `services.ProxyWatcher` and `services.Fanout`.  Simple self-contained helpers which are technically application logic, but basically just wrap the interfaces defined in `services` and make them a bit easier to use in certain ways.  Maybe we should create something like a `services/helpers` for things like this to live in?

Reply:
Okay, let's pass on the interface then.